File: go/utils/price_updater.go
package utils

import (
	"context"
	"encoding/json"
	"log"

	"github.com/go-redis/redis/v8"
	"github.com/soulmachine/coinsignal/config"
	"github.com/soulmachine/coinsignal/pojo"
)

type PriceUpdater struct {
	pubsub *redis.PubSub
	prices map[string]float64
}

func NewPriceUpdater(ctx context.Context, redis_url string) *PriceUpdater {
	rdb := redis.NewClient(&redis.Options{
		Addr: redis_url,
	})
	pubsub := rdb.Subscribe(ctx, config.REDIS_TOPIC_MARK_PRICE)

	updater := &PriceUpdater{
		pubsub: pubsub,
		prices: make(map[string]float64),
	}

	go updater.run()

	return updater
}

func (updater *PriceUpdater) GetPrice(currency string) float64 {
	price := updater.prices[currency]
	return price
}

func (updater *PriceUpdater) Close() {
	updater.pubsub.Close()
}

func (updater *PriceUpdater) run() {
	ch := updater.pubsub.Channel()
	// Consume messages.
	for msg := range ch {
		var mark_price pojo.MarkPrice
		if err := json.Unmarshal([]byte(msg.Payload), &mark_price); err != nil {
			log.Fatalln(err)
		} else {
			updater.prices[mark_price.Currency] = mark_price.Price
		}
	}
}

File: rust/utils/src/price_updater.rs
use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
    thread,
};

pub struct PriceUpdater {
    redis_url: String,
    prices: Arc<Mutex<HashMap<String, f64>>>,
}

#[derive(Serialize, Deserialize)]
struct MarkPrice {
    currency: String,
    price: f64,
}

impl PriceUpdater {
    pub fn new(redis_url: &str) -> Self {
        let updater = PriceUpdater {
            redis_url: redis_url.to_string(),
            prices: Arc::new(Mutex::new(HashMap::new())),
        };

        updater.run(); // create a thread

        updater
    }

    pub fn get_price(&self, currency: &str) -> Option<f64> {
        if let Some(price) = self.prices.lock().unwrap().get(currency) {
            Some(*price)
        } else {
            None
        }
    }

    fn run(&self) {
        let prices_clone = self.prices.clone();
        let redis_url = self.redis_url.to_string();
        thread::spawn(move || {
            let client = redis::Client::open(redis_url).unwrap();
            let mut connection = client.get_connection().unwrap();
            let mut pubsub = connection.as_pubsub();
            pubsub.subscribe("coinsignal:mark_price").unwrap();

            loop {
                let msg = pubsub.get_message().unwrap();
                let payload: String = msg.get_payload().unwrap();
                if let Ok(mark_price) = serde_json::from_str::<MarkPrice>(&payload) {
                    let mut guard = prices_clone.lock().unwrap();
                    guard.insert(mark_price.currency.clone(), mark_price.price);
                }
            }
        });
    }
}
